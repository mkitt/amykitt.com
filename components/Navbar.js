// @flow
import React from 'react'
import { createSelector } from 'reselect'
import css, { media3 } from '../styles/css'
import {
  addResizeObserver,
  addScrollObserver,
  getPageY,
  removeResizeObserver,
  removeScrollObserver,
  scrollViewport,
} from '../lib/viewport'
import NavbarBullet from './NavbarBullet'
import NavbarLink from './NavbarLink'
import NavbarLogo from './NavbarLogo'
import NavbarTitle from './NavbarTitle'
import View from './View'
import type { ResizeProps, ScrollProps } from '../types/app.js.flow'

type Props = {
  pathname: string,
}

type State = {
  isAboutActive: boolean,
  viewportColumnCount: number,
  viewportHeight: number,
}

const navStyle = css({
  position: 'sticky',
  top: 0,
  zIndex: 5,
  height: 160,
  color: '#7c7c7c',
  backgroundColor: '#f6f6f6',
})

const wrapperStyle = css(
  {
    position: 'relative',
    display: 'flex',
    flexFlow: 'row wrap',
    alignItems: 'center',
    justifyContent: 'center',
    maxWidth: 1440,
    height: '100%',
    marginRight: 'auto',
    marginLeft: 'auto',
    paddingRight: 20,
    paddingLeft: 20,
    outline: '1px dotted magenta',
  },
  media3({
    paddingRight: 80,
    paddingLeft: 80,
  }),
)

const contentStyle = css({
  flex: 2,
  display: 'flex',
  flexFlow: 'column wrap',
  alignItems: 'center',
  justifyContent: 'center',
})

const selectViewportHeight = state => state.viewportHeight || 0

const selectAboutY = createSelector(
  [selectViewportHeight], (height) => {
    const element = document.getElementById('about')
    return element && height ? (element.offsetTop - 160) : NaN
  },
)

const onClickToSection = (e: any) => {
  e.preventDefault()
  const href = e.target.getAttribute('href')
  const selector = href.replace('#', '')
  if (selector === '/') {
    scrollViewport(0)
    return
  }
  const element = document.getElementById(selector)
  const top = element && (element.getBoundingClientRect().top + getPageY()) - 160
  scrollViewport(top || 0)
}

export default class extends React.PureComponent {
  props: Props
  state: State = {
    isAboutActive: false,
    viewportColumnCount: 5,
    viewportHeight: 768,
  }

  componentDidMount() {
    addResizeObserver(this)
    addScrollObserver(this)
  }

  componentWillUnmount() {
    removeResizeObserver(this)
    removeScrollObserver(this)
  }

  onResize = ({ viewportColumnCount, viewportHeight }: ResizeProps) => {
    this.setState({ viewportColumnCount, viewportHeight })
  }

  onScroll = (props: ScrollProps) => {
    const aboutY = selectAboutY(this.state)
    const isAboutActive = !isNaN(aboutY) && props.scrollY >= aboutY
    if (this.state.isAboutActive !== isAboutActive) {
      this.setState({ isAboutActive })
    }
  }

  render() {
    const { isAboutActive, viewportColumnCount } = this.state
    return (
      <nav className={navStyle}>
        <View className={wrapperStyle}>
          { viewportColumnCount > 2 &&
            <NavbarLogo
              href="/"
              isActive={!isAboutActive}
              onClick={onClickToSection}
            />
          }
          <View className={contentStyle}>
            <NavbarTitle />
            <View>
              <NavbarLink
                href="/"
                isActive={!isAboutActive}
                onClick={onClickToSection}
              >
                Work
              </NavbarLink>
              <NavbarBullet />
              <NavbarLink
                href="#about"
                isActive={isAboutActive}
                onClick={onClickToSection}
              >
                About
              </NavbarLink>
            </View>
          </View>
        </View>
      </nav>
    )
  }
}
