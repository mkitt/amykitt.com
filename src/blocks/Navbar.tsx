import React, { memo, SyntheticEvent } from 'react'
import Mark from '../assets/Mark'
import Container from '../elements/Container'
import Flex from '../elements/Flex'
import Grid from '../elements/Grid'
import Link from '../elements/Link'
import Text from '../elements/Text'
import View from '../elements/View'

const getLinkStyle = ({ isActive }: { isActive?: boolean }) => ({
  borderBottom: `1px solid ${
    isActive === true ? 'currentColor' : 'transparent'
  }`,
  color: isActive === true ? '#000' : 'currentColor',
  transition: 'color 0.333s',
  '.no-touch &:hover': {
    color: '#000',
  },
})

type PropsType = {
  handleClickToSection: (e: SyntheticEvent) => void
  isAboutActive: boolean
}

const Navbar = ({ handleClickToSection, isAboutActive }: PropsType) => (
  <Flex
    as="nav"
    css={{
      alignItems: 'center',
      backgroundColor: '#fff',
      color: '#7c7c7c',
      fontSize: '0.75rem',
      fontWeight: 700,
      height: '10rem',
      position: 'sticky',
      textTransform: 'uppercase',
      top: 0,
      zIndex: 5,
    }}
    role="navigation"
  >
    <Container
      css={{
        '@media (min-width: 47.5em)': {
          gridTemplateColumns: '5rem 1fr 5rem',
        },
      }}
    >
      <View
        css={{
          '@media (max-width: 47.4375em)': {
            display: 'none',
          },
        }}
      >
        <Link href="/" onClick={handleClickToSection} title="AK Design">
          <Mark />
        </Link>
      </View>
      <Grid css={{ alignItems: 'center', justifyContent: 'center' }}>
        <Text
          as="h1"
          css={{
            fontSize: 'inherit',
            fontWeight: 400,
            margin: 0,
            textTransform: 'uppercase',
          }}
        >
          <strong>Hi. </strong>
          <span>
            If you like what you see below perhaps we can work together in the
            future.
          </span>
        </Text>
        <Flex
          css={{
            alignItems: 'center',
            justifyContent: 'center',
          }}
        >
          <Link
            css={getLinkStyle({ isActive: !isAboutActive })}
            href="/"
            onClick={handleClickToSection}
            title="View Work"
          >
            Work
          </Link>
          <Text css={{ padding: '0 0.5rem' }}>&bull;</Text>
          <Link
            css={getLinkStyle({ isActive: isAboutActive })}
            href="#about"
            onClick={handleClickToSection}
            title={`About AK Design ${isAboutActive}`}
          >
            About
          </Link>
          <Text css={{ padding: '0 0.5rem' }}>&bull;</Text>
          <Link
            css={getLinkStyle({ isActive: false })}
            href="mailto:amy.kitt@gmail.com"
            title="Say hello!"
          >
            Contact
          </Link>
        </Flex>
      </Grid>
    </Container>
  </Flex>
)

export default memo(Navbar)
