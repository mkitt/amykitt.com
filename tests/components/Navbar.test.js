import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Navbar from '../../components/Navbar'

describe('Navbar', () => {
  let nav

  beforeEach(() => {
    nav = <Navbar pathname="/" />
  })

  it('sets up the correct navigation', () => {
    expect(shallow(nav).find('a').at(0).text()).toEqual('<RocketIcon />')
    expect(shallow(nav).find('a').at(1).text()).toEqual('Simple')
    expect(shallow(nav).find('a').at(2).text()).toEqual('GitHub')
    expect(shallow(nav).find('a').at(3).text()).toEqual('Navless')
  })

  it('sets the correct active state (Home)', () => {
    const active = shallow(nav).find('a').at(0).props().className
    const inactive = shallow(nav).find('a').at(1).props().className
    const inactive2 = shallow(nav).find('a').at(1).props().className
    expect(active).not.toEqual(inactive)
    expect(inactive).toEqual(inactive2)
  })

  it('shows the correct snapshot', () => {
    const component = renderer.create(nav)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

