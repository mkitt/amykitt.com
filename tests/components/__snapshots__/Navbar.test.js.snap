// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar shows the correct snapshot for the Navbar component tree 1`] = `
<nav
  className={
    Object {
      "data-css-1yiesa8": "",
    }
  }
>
  <div
    className={
      Object {
        "data-css-1abfnkv": "",
      }
    }
  >
    <a
      className="css-19avx8h"
      href="/"
      onClick={[Function]}
      title="AK Design"
    >
      <svg
        aria-hidden={true}
        height={47}
        viewBox="0 0 73 80"
        width={43}
      >
        <g>
          <path
            d="M0,37.4468085 L0,29.5421277 L7.6313703,29.5421277 C19.0775814,28.9514894 25.2042934,24.9957447 25.2042934,8.90893617 L25.2042934,0 L33.7745975,0 L33.7745975,37.4468085 L0,37.4468085 Z"
            fill="#7c7c7c"
          />
          <path
            d="M33.7745975,80 L33.7745975,72.0953191 L26.1432272,72.0953191 C14.6970161,71.506383 8.57030411,67.5506383 8.57030411,51.4621277 L8.57030411,42.5531915 L0,42.5531915 L0,80 L33.7745975,80 Z"
            fill="#7c7c7c"
          />
          <path
            d="M72.6153846,37.4468085 L72.6153846,29.5421277 L64.9840143,29.5421277 C53.5378032,28.9514894 47.4110912,24.9957447 47.4110912,8.90893617 L47.4110912,0 L38.8407871,0 L38.8407871,37.4468085 L72.6153846,37.4468085 Z"
            fill="#e1aee8"
          />
        </g>
      </svg>
    </a>
    <div
      className={
        Object {
          "data-css-u3lxy3": "",
        }
      }
    >
      <h1
        className={
          Object {
            "data-css-1a2lm7n": "",
          }
        }
      >
        <span
          className={
            Object {
              "data-css-41sc57": "",
            }
          }
        >
          Hi. 
        </span>
        <span
          className={
            Object {
              "data-css-ucwin8": "",
            }
          }
        >
          If you like what you see below perhaps we can work together in the future.
        </span>
      </h1>
      <div>
        <a
          className="css-e3p90y css-863xba"
          href="/"
          onClick={[Function]}
        >
          Work
        </a>
        <span
          className={
            Object {
              "data-css-3eixqd": "",
            }
          }
        >
          â€¢
        </span>
        <a
          className="css-e3p90y "
          href="#about"
          onClick={[Function]}
        >
          About
        </a>
      </div>
    </div>
  </div>
</nav>
`;
